shell/platform/windows/flutter_windows_engine.cc

```diff
@@ -492,34 +494,118 @@ bool FlutterWindowsEngine::Stop() {
+    // https://github.com/flutter/flutter/issues/146248
+    captures.latch.Wait();
+
+    if (!captures.added) {
+      // Adding the view failed. Update the embedder's state to match the
+      // engine's state. This is unexpected and indicates a bug in the Windows
+      // embedder.
+      FML_LOG(ERROR) << "FlutterEngineAddView failed to add view";
+      fml::UniqueLock write_lock{*views_mutex_};
+      views_.erase(view_id);
+      return nullptr;
+    }
+  }
 
   return std::move(view);
 }
 
 void FlutterWindowsEngine::RemoveView(FlutterViewId view_id) {
   FML_DCHECK(running());
-  FML_DCHECK(views_.find(view_id) != views_.end());
 
-  if (view_id == kImplicitViewId) {
-    // The engine and framework assume the implicit view always exists.
-    // Attempts to render to the implicit view will be ignored.
-    views_.erase(view_id);
-    return;
+  // Notify the engine to stop rendering to the view if it isn't the implicit
+  // view. The engine and framework assume the implicit view always exists and
+  // can continue presenting.
+  if (view_id != kImplicitViewId) {
+    struct Captures {
+      fml::AutoResetWaitableEvent latch;
+      bool removed;
+    };
+    Captures captures;
+
+    FlutterRemoveViewInfo info = {};
+    info.struct_size = sizeof(FlutterRemoveViewInfo);
+    info.view_id = view_id;
+    info.user_data = &captures;
+    info.remove_view_callback = [](const FlutterRemoveViewResult* result) {
+      // This is invoked on the raster thread, the same thread that the present
+      // callback is invoked. If |FlutterRemoveViewResult.removed| is `true`,
+      // the engine guarantees the view won't be presented.
+      Captures* captures = reinterpret_cast<Captures*>(result->user_data);
+      captures->removed = result->removed;
+      captures->latch.Signal();
+    };
+
+    embedder_api_.RemoveView(engine_, &info);
+
+    // Block the platform thread until the engine has removed the view.
+    // TODO(loicsharma): This blocks the platform thread eagerly and can
+    // cause unnecessary delay in input processing. Instead, this should block
+    // lazily only when an operation needs the view.
+    // https://github.com/flutter/flutter/issues/146248
+    captures.latch.Wait();
+
+    if (!captures.removed) {
+      // Removing the view failed. This is unexpected and indicates a bug in the
+      // Windows embedder.
+      FML_LOG(ERROR) << "FlutterEngineRemoveView failed to remove view";
+      return;
+    }
   }
 
-  // TODO(loicsharma): Remove the view from the engine using the
-  // `FlutterEngineRemoveView` embedder API. Windows does not
-  // support views other than the implicit view yet.
-  // https://github.com/flutter/flutter/issues/144810
-  FML_UNREACHABLE();
+  {
+    // The engine no longer presents to the view. Remove the view from the
+    // embedder.
+    fml::UniqueLock write_lock{*views_mutex_};
+
+    FML_DCHECK(views_.find(view_id) != views_.end());
```

Comment:
We should probably not assert on user error. Maybe return a bool to indicate if the removal was successful.

Reply:
This indicates a bug in the embedder, not a user error.

The user cannot destroy views by ID manually. Instead, a user creates and destroys `FlutterViewController`s. The controller's destructor eventually results in a call to `FlutterWindowsEngine::RemoveView` with the view's ID.

Reply:
I see. If the view ID is invalid it should have already returned as indicated by `captures.removed`.